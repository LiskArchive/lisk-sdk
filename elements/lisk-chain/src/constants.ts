/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { hash } from '@liskhq/lisk-cryptography';

export const CHAIN_STATE_BURNT_FEE = 'burntFee';
export const CONSENSUS_STATE_FINALIZED_HEIGHT_KEY = 'finalizedHeight';
export const CONSENSUS_STATE_VALIDATORS_KEY = 'validators';

export const DEFAULT_MIN_BLOCK_HEADER_CACHE = 309;
export const DEFAULT_MAX_BLOCK_HEADER_CACHE = 515;

export const EVENT_NEW_BLOCK = 'EVENT_NEW_BLOCK';
export const EVENT_DELETE_BLOCK = 'EVENT_DELETE_BLOCK';

export const EMPTY_BUFFER = Buffer.alloc(0);
export const EMPTY_HASH = hash(EMPTY_BUFFER);

export const GENESIS_BLOCK_VERSION = 0;
export const GENESIS_BLOCK_GENERATOR_PUBLIC_KEY = EMPTY_BUFFER;
export const GENESIS_BLOCK_REWARD = BigInt(0);
export const GENESIS_BLOCK_SIGNATURE = EMPTY_BUFFER;
export const GENESIS_BLOCK_TRANSACTION_ROOT = EMPTY_HASH;
