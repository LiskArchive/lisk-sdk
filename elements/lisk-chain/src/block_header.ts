/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { ed, utils } from '@liskhq/lisk-cryptography';
import { codec } from '@liskhq/lisk-codec';
import { validator, LiskValidationError } from '@liskhq/lisk-validator';
import { LiskErrorObject } from '@liskhq/lisk-validator/dist-node/types';
import { EMPTY_BUFFER, EMPTY_HASH, SIGNATURE_LENGTH_BYTES, TAG_BLOCK_HEADER } from './constants';
import { blockHeaderSchema, blockHeaderSchemaWithId, signingBlockHeaderSchema } from './schema';
import { JSONObject } from './types';

export interface BlockHeaderAttrs {
	readonly version: number;
	readonly height: number;
	readonly generatorAddress: Buffer;
	readonly previousBlockID: Buffer;
	readonly timestamp: number;
	readonly maxHeightPrevoted: number;
	readonly maxHeightGenerated: number;
	readonly impliesMaxPrevote: boolean;
	readonly aggregateCommit: {
		readonly height: number;
		readonly aggregationBits: Buffer;
		readonly certificateSignature: Buffer;
	};
	readonly validatorsHash?: Buffer;
	readonly stateRoot?: Buffer;
	readonly transactionRoot?: Buffer;
	readonly assetRoot?: Buffer;
	readonly eventRoot?: Buffer;
	signature?: Buffer;
	id?: Buffer;
}

export type BlockHeaderJSON = JSONObject<BlockHeaderAttrs>;

export class BlockHeader {
	public readonly version: number;
	public readonly height: number;
	public readonly generatorAddress: Buffer;
	public readonly previousBlockID: Buffer;
	public readonly timestamp: number;
	public readonly maxHeightPrevoted: number;
	public readonly maxHeightGenerated: number;
	public readonly impliesMaxPrevote: boolean;
	private _aggregateCommit: {
		height: number;
		aggregationBits: Buffer;
		certificateSignature: Buffer;
	};
	private _validatorsHash?: Buffer;
	private _stateRoot?: Buffer;
	private _transactionRoot?: Buffer;
	private _assetRoot?: Buffer;
	private _eventRoot?: Buffer;
	private _signature?: Buffer;
	private _id?: Buffer;

	public constructor({
		version,
		timestamp,
		height,
		generatorAddress,
		previousBlockID,
		maxHeightPrevoted,
		maxHeightGenerated,
		impliesMaxPrevote,
		aggregateCommit,
		validatorsHash,
		stateRoot,
		eventRoot,
		assetRoot,
		transactionRoot,
		signature,
		id,
	}: BlockHeaderAttrs) {
		this.version = version;
		this.height = height;
		this.generatorAddress = generatorAddress;
		this.previousBlockID = previousBlockID;
		this.timestamp = timestamp;
		this.maxHeightPrevoted = maxHeightPrevoted;
		this.maxHeightGenerated = maxHeightGenerated;
		this.impliesMaxPrevote = impliesMaxPrevote;
		this._aggregateCommit = aggregateCommit;
		this._validatorsHash = validatorsHash;
		this._eventRoot = eventRoot;
		this._stateRoot = stateRoot;
		this._transactionRoot = transactionRoot;
		this._assetRoot = assetRoot;

		this._signature = signature;
		this._id = id;
	}

	public static fromBytes(value: Buffer): BlockHeader {
		return new BlockHeader(codec.decode<BlockHeaderAttrs>(blockHeaderSchema, value));
	}

	public static fromJSON(value: Record<string, unknown>): BlockHeader {
		return new BlockHeader(codec.fromJSON<BlockHeaderAttrs>(blockHeaderSchema, value));
	}

	public get stateRoot() {
		return this._stateRoot;
	}

	public set stateRoot(val) {
		this._stateRoot = val;
		this._resetComputedValues();
	}

	public get eventRoot() {
		return this._eventRoot;
	}

	public set eventRoot(val) {
		this._eventRoot = val;
		this._resetComputedValues();
	}

	public get assetRoot() {
		return this._assetRoot;
	}

	public set assetRoot(val) {
		this._assetRoot = val;
		this._resetComputedValues();
	}

	public get transactionRoot() {
		return this._transactionRoot;
	}

	public set transactionRoot(val) {
		this._transactionRoot = val;
		this._resetComputedValues();
	}

	public get validatorsHash() {
		return this._validatorsHash;
	}

	public set validatorsHash(val) {
		this._validatorsHash = val;
		this._resetComputedValues();
	}

	public get aggregateCommit() {
		return this._aggregateCommit;
	}

	public set aggregateCommit(val) {
		this._aggregateCommit = val;
		this._resetComputedValues();
	}

	public getBytes(): Buffer {
		return codec.encode(blockHeaderSchema, this._getBlockHeaderProps());
	}

	public toJSON(): BlockHeaderJSON {
		return codec.toJSON(blockHeaderSchemaWithId, this._getAllProps());
	}

	public toObject(): Required<BlockHeaderAttrs> {
		return this._getAllProps();
	}

	public validate(): void {
		validator.validate(blockHeaderSchema, this._getBlockHeaderProps());

		if (this.previousBlockID.length === 0) {
			throw new Error('Previous block id must not be empty.');
		}
		if (this.signature.length !== SIGNATURE_LENGTH_BYTES) {
			throw new Error('Signature length must be 64 bytes.');
		}
	}

	public validateGenesis(): void {
		const header = this._getBlockHeaderProps();
		validator.validate(blockHeaderSchema, header);

		const errors: LiskErrorObject[] = [];

		if (header.previousBlockID.length !== 32) {
			errors.push({
				message: 'Genesis block header previousBlockID must be 32 bytes',
				keyword: 'const',
				dataPath: 'header.previousBlockID',
				schemaPath: 'properties.previousBlockID',
				params: {},
			});
		}

		if (!header.transactionRoot.equals(EMPTY_HASH)) {
			errors.push({
				message: 'Genesis block header transaction root must be empty hash',
				keyword: 'const',
				dataPath: 'header.transactionRoot',
				schemaPath: 'properties.transactionRoot',
				params: { allowedValue: EMPTY_HASH },
			});
		}

		if (header.maxHeightPrevoted !== header.height) {
			errors.push({
				message: 'Genesis block header maxHeightPrevoted must equal height',
				keyword: 'const',
				dataPath: 'header.maxHeightPrevoted',
				schemaPath: 'properties.maxHeightPrevoted',
				params: { allowedValue: header.height },
			});
		}

		if (header.maxHeightGenerated !== 0) {
			errors.push({
				message: 'Genesis block header maxHeightGenerated must equal 0',
				keyword: 'const',
				dataPath: 'header.maxHeightGenerated',
				schemaPath: 'properties.maxHeightGenerated',
				params: { allowedValue: 0 },
			});
		}

		if (!header.impliesMaxPrevote) {
			errors.push({
				message: 'Genesis block header impliesMaxPrevote must be true',
				keyword: 'const',
				dataPath: 'header.impliesMaxPrevote',
				schemaPath: 'properties.impliesMaxPrevote',
				params: { allowedValue: true },
			});
		}

		if (header.aggregateCommit.height !== 0) {
			errors.push({
				message: 'Genesis block header aggregateCommit.height must equal 0',
				keyword: 'const',
				dataPath: 'aggregateCommit.height',
				schemaPath: 'properties.aggregateCommit.height',
				params: { allowedValue: 0 },
			});
		}

		if (!header.aggregateCommit.certificateSignature.equals(EMPTY_BUFFER)) {
			errors.push({
				message: 'Genesis block header aggregateCommit.certificateSignature must be empty bytes',
				keyword: 'const',
				dataPath: 'aggregateCommit.certificateSignature',
				schemaPath: 'properties.aggregateCommit.certificateSignature',
				params: { allowedValue: EMPTY_BUFFER },
			});
		}

		if (!header.aggregateCommit.aggregationBits.equals(EMPTY_BUFFER)) {
			errors.push({
				message: 'Genesis block header aggregateCommit.aggregationBits must be empty bytes',
				keyword: 'const',
				dataPath: 'aggregateCommit.aggregationBits',
				schemaPath: 'properties.aggregateCommit.aggregationBits',
				params: { allowedValue: EMPTY_BUFFER },
			});
		}

		if (!header.signature.equals(EMPTY_BUFFER)) {
			errors.push({
				message: 'Genesis block header signature must be empty bytes',
				keyword: 'const',
				dataPath: 'header.signature',
				schemaPath: 'properties.signature',
				params: { allowedValue: EMPTY_BUFFER },
			});
		}

		if (errors.length) {
			throw new LiskValidationError(errors);
		}
	}

	public validateSignature(publicKey: Buffer, chainID: Buffer): void {
		const signingBytes = this.getSigningBytes();
		const valid = ed.verifyData(TAG_BLOCK_HEADER, chainID, signingBytes, this.signature, publicKey);

		if (!valid) {
			throw new Error('Invalid block signature.');
		}
	}

	public getSigningBytes(): Buffer {
		const blockHeaderBytes = codec.encode(signingBlockHeaderSchema, this._getSigningProps());

		return blockHeaderBytes;
	}

	public sign(chainID: Buffer, privateKey: Buffer): void {
		this._signature = ed.signDataWithPrivateKey(
			TAG_BLOCK_HEADER,
			chainID,
			this.getSigningBytes(),
			privateKey,
		);
	}

	public get signature(): Buffer {
		if (!this._signature) {
			throw new Error('Block header is not signed.');
		}

		return this._signature;
	}

	public get id(): Buffer {
		if (!this._id && !this._signature) {
			throw new Error('Can not generate the id for unsigned block header.');
		}

		if (!this._id) {
			const blockHeaderBytes = codec.encode(blockHeaderSchema, {
				...this._getSigningProps(),
				signature: this._signature,
			});

			this._id = utils.hash(blockHeaderBytes);
		}

		return this._id;
	}

	private _resetComputedValues() {
		this._id = undefined;
	}

	private _getSigningProps() {
		if (!this.assetRoot) {
			throw new Error('Asset root is empty.');
		}
		if (!this.eventRoot) {
			throw new Error('Event root is empty.');
		}
		if (!this.stateRoot) {
			throw new Error('State root is empty.');
		}
		if (!this.transactionRoot) {
			throw new Error('Transaction root is empty.');
		}
		if (!this.validatorsHash) {
			throw new Error('Validators hash is empty.');
		}
		return {
			version: this.version,
			timestamp: this.timestamp,
			height: this.height,
			previousBlockID: this.previousBlockID,
			stateRoot: this.stateRoot,
			assetRoot: this.assetRoot,
			eventRoot: this.eventRoot,
			transactionRoot: this.transactionRoot,
			validatorsHash: this.validatorsHash,
			aggregateCommit: this.aggregateCommit,
			generatorAddress: this.generatorAddress,
			maxHeightPrevoted: this.maxHeightPrevoted,
			maxHeightGenerated: this.maxHeightGenerated,
			impliesMaxPrevote: this.impliesMaxPrevote,
		};
	}

	private _getBlockHeaderProps() {
		return {
			...this._getSigningProps(),
			signature: this.signature,
		};
	}

	private _getAllProps() {
		return {
			...this._getBlockHeaderProps(),
			id: this.id,
		};
	}
}
