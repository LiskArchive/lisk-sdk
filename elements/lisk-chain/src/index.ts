/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import { EVENT_DELETE_BLOCK, EVENT_NEW_BLOCK } from './constants';

const events = { EVENT_DELETE_BLOCK, EVENT_NEW_BLOCK };

export { events };
export { Chain } from './chain';
export { Transaction, transactionSchema } from './transaction';
export {
	blockHeaderAssetSchema,
	blockHeaderSchema,
	blockSchema,
	signingBlockHeaderSchema,
	validatorsSchema,
} from './schema';
export type {
	Account,
	AccountDefaultProps,
	RawBlock,
	RawBlockHeader,
	GenesisBlock,
	GenesisBlockHeader,
	Block,
	BlockHeader,
	Validator,
	AccountSchema,
} from './types';
export { readGenesisBlockJSON } from './utils';

export type { StateStore } from './state_store';
export type { Slots } from './slots';
