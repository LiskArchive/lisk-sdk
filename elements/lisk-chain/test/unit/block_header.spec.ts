/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import { utils } from '@liskhq/lisk-cryptography';
import { BlockHeader } from '../../src/block_header';
import { EMPTY_BUFFER, EMPTY_HASH } from '../../src/constants';
import {
	blockHeaderSchema,
	blockHeaderSchemaWithId,
	signingBlockHeaderSchema,
} from '../../src/schema';

const getBlockAttrs = () => ({
	version: 1,
	timestamp: 1009988,
	height: 1009988,
	previousBlockID: Buffer.from('4a462ea57a8c9f72d866c09770e5ec70cef18727', 'hex'),
	stateRoot: Buffer.from('7f9d96a09a3fd17f3478eb7bef3a8bda00e1238b', 'hex'),
	transactionRoot: Buffer.from('b27ca21f40d44113c2090ca8f05fb706c54e87dd', 'hex'),
	assetRoot: Buffer.from('b27ca21f40d44113c2090ca8f05fb706c54e87dd', 'hex'),
	eventRoot: Buffer.from('30dda4fbc395828e5a9f2f8824771e434fce4945a1e7820012440d09dd1e2b6d', 'hex'),
	generatorAddress: Buffer.from('be63fb1c0426573352556f18b21efd5b6183c39c', 'hex'),
	maxHeightPrevoted: 1000988,
	maxHeightGenerated: 1000988,
	impliesMaxPrevotes: true,
	validatorsHash: utils.hash(Buffer.alloc(0)),
	aggregateCommit: {
		height: 0,
		aggregationBits: Buffer.alloc(0),
		certificateSignature: Buffer.alloc(0),
	},
	signature: Buffer.from('6da88e2fd4435e26e02682435f108002ccc3ddd5', 'hex'),
});

const getGenesisBlockAttrs = () => ({
	version: 1,
	timestamp: 1009988,
	height: 1009988,
	previousBlockID: utils.getRandomBytes(32),
	stateRoot: Buffer.from('7f9d96a09a3fd17f3478eb7bef3a8bda00e1238b', 'hex'),
	transactionRoot: EMPTY_HASH,
	assetRoot: EMPTY_HASH,
	eventRoot: EMPTY_HASH,
	impliesMaxPrevotes: true,
	generatorAddress: Buffer.alloc(20, 0),
	maxHeightPrevoted: 1009988,
	maxHeightGenerated: 0,
	validatorsHash: utils.hash(Buffer.alloc(0)),
	aggregateCommit: {
		height: 0,
		aggregationBits: Buffer.alloc(0),
		certificateSignature: EMPTY_BUFFER,
	},
	signature: EMPTY_BUFFER,
});

const blockId = Buffer.from(
	'926ef9e8f01294c9b1f068119002a3903f5a2820b957759402faefeb8fd6e97a',
	'hex',
);

const blockHeaderBytes = Buffer.from(
	'080110c4d23d18c4d23d22144a462ea57a8c9f72d866c09770e5ec70cef187272a14be63fb1c0426573352556f18b21efd5b6183c39c3214b27ca21f40d44113c2090ca8f05fb706c54e87dd3a14b27ca21f40d44113c2090ca8f05fb706c54e87dd422030dda4fbc395828e5a9f2f8824771e434fce4945a1e7820012440d09dd1e2b6d4a147f9d96a09a3fd17f3478eb7bef3a8bda00e1238b509c8c3d589c8c3d60016a20e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b8557206080012001a007a146da88e2fd4435e26e02682435f108002ccc3ddd5',
	'hex',
);

const blockHeaderProps = [
	'version',
	'timestamp',
	'height',
	'previousBlockID',
	'generatorAddress',
	'transactionRoot',
	'assetRoot',
	'stateRoot',
	'maxHeightPrevoted',
	'maxHeightGenerated',
	'validatorsHash',
	'aggregateCommit',
];

describe('block_header', () => {
	describe('signingBlockHeaderSchema', () => {
		it('should be valid schema without signature and id', () => {
			expect(signingBlockHeaderSchema).toMatchSnapshot();
			expect(Object.keys(signingBlockHeaderSchema.properties)).toIncludeAllMembers([
				...blockHeaderProps,
			]);
			expect(Object.keys(signingBlockHeaderSchema.properties)).not.toIncludeAllMembers([
				'id',
				'signature',
			]);
		});
	});

	describe('blockHeaderSchema', () => {
		it('should be valid schema with signature but without id', () => {
			expect(blockHeaderSchema).toMatchSnapshot();
			expect(Object.keys(blockHeaderSchema.properties)).toIncludeAllMembers([
				...blockHeaderProps,
				'signature',
			]);
			expect(Object.keys(blockHeaderSchema.properties)).not.toIncludeAllMembers(['id']);
		});
	});

	describe('blockHeaderSchemaWithId', () => {
		it('should be valid schema with signature and id', () => {
			expect(blockHeaderSchemaWithId).toMatchSnapshot();
			expect(Object.keys(blockHeaderSchemaWithId.properties)).toIncludeAllMembers([
				...blockHeaderProps,
				'signature',
				'id',
			]);
		});
	});

	describe('BlockHeader', () => {
		describe('constructor', () => {
			it('should initialize block header object', () => {
				const data = getBlockAttrs();

				const blockHeader = new BlockHeader(data);

				expect(blockHeader).toBeInstanceOf(BlockHeader);
				expect(blockHeader.version).toEqual(data.version);
				expect(blockHeader.timestamp).toEqual(data.timestamp);
				expect(blockHeader.height).toEqual(data.height);
				expect(blockHeader.generatorAddress).toEqual(data.generatorAddress);
				expect(blockHeader.previousBlockID).toEqual(data.previousBlockID);
				expect(blockHeader.stateRoot).toEqual(data.stateRoot);
				expect(blockHeader.eventRoot).toEqual(data.eventRoot);
				expect(blockHeader.validatorsHash).toEqual(data.validatorsHash);
				expect(blockHeader.aggregateCommit).toEqual(data.aggregateCommit);
				expect(blockHeader.maxHeightPrevoted).toEqual(data.maxHeightPrevoted);
				expect(blockHeader.maxHeightGenerated).toEqual(data.maxHeightGenerated);
				expect(blockHeader.assetRoot).toEqual(data.assetRoot);
				expect(blockHeader.transactionRoot).toEqual(data.transactionRoot);
				expect(blockHeader.id).toEqual(blockId);
				expect(blockHeader.signature).toEqual(data.signature);
			});
		});

		describe('fromBytes', () => {
			it('should load the block header from bytes', () => {
				const blockHeader = BlockHeader.fromBytes(blockHeaderBytes);

				expect(blockHeader.toObject()).toEqual({ ...getBlockAttrs(), id: blockId });
			});
		});

		describe('validate', () => {
			it('should validate the block header without error if all the properties are valid', () => {
				const data = getBlockAttrs();
				data.signature = Buffer.from(
					'41f7d923c8957664923b49d7a893153476bc60e0392702cf111a26a92d74279a00c41f1999504c9faeee7b3d05393ca04d61a0768c3ae4d324f3097ab0b52201',
					'hex',
				);
				const blockHeader = new BlockHeader(data);

				expect(blockHeader.validate()).toBeUndefined();
			});

			it('should throw error if previous block id is not set', () => {
				const data = getBlockAttrs();
				const blockHeader = new BlockHeader({ ...data, previousBlockID: Buffer.alloc(0) });

				expect(() => blockHeader.validate()).toThrow('Previous block id must not be empty.');
			});

			it('should throw error if signature length is not correct', () => {
				const data = getBlockAttrs();
				const blockHeader = new BlockHeader(data);

				expect(() => blockHeader.validate()).toThrow('Signature length must be 64 bytes.');
			});
		});

		describe('signature', () => {
			it('should throw error if block header not signed', () => {
				const { signature, ...rest } = getBlockAttrs();
				const blockHeader = new BlockHeader(rest);

				expect(() => blockHeader.signature).toThrow('Block header is not signed.');
			});
		});

		describe('id', () => {
			it('should throw error if block header not signed', () => {
				const { signature, ...rest } = getBlockAttrs();
				const blockHeader = new BlockHeader(rest);

				expect(() => blockHeader.id).toThrow('Can not generate the id for unsigned block header.');
			});

			it('should not throw error if block header is signed', () => {
				const data = getBlockAttrs();
				const blockHeader = new BlockHeader(data);

				expect(blockHeader.id).toEqual(blockId);
			});
		});

		describe('validateGenesis', () => {
			it('should throw error if previousBlockID is not 32 bytes', () => {
				const block = getGenesisBlockAttrs();
				const blockHeader = new BlockHeader({
					...block,
					previousBlockID: utils.getRandomBytes(31),
				});

				expect(() => blockHeader.validateGenesis()).toThrow(
					'Genesis block header previousBlockID must be 32 bytes',
				);
			});

			it('should throw error if generatorAddress is zero bytes', () => {
				const block = getGenesisBlockAttrs();
				const blockHeader = new BlockHeader({
					...block,
					generatorAddress: utils.getRandomBytes(20),
				});

				expect(() => blockHeader.validateGenesis()).toThrow(
					'Genesis block header generator address must be empty bytes',
				);
			});

			it('should throw error if transactionRoot is not empty hash', () => {
				const block = getGenesisBlockAttrs();
				const blockHeader = new BlockHeader({
					...block,
					transactionRoot: utils.getRandomBytes(32),
				});

				expect(() => blockHeader.validateGenesis()).toThrow(
					'Genesis block header transaction root must be empty hash',
				);
			});

			it('should throw error if maxHeightPrevoted is not equal to header.height', () => {
				const block = getGenesisBlockAttrs();
				const blockHeader = new BlockHeader({ ...block, maxHeightPrevoted: 10 });

				expect(() => blockHeader.validateGenesis()).toThrow(
					'Genesis block header maxHeightPrevoted must equal height',
				);
			});

			it('should throw error if aggregateCommit.height is not equal to 0', () => {
				const block = getGenesisBlockAttrs();
				const blockHeader = new BlockHeader({
					...block,
					aggregateCommit: { ...block.aggregateCommit, height: 10 },
				});

				expect(() => blockHeader.validateGenesis()).toThrow(
					'Genesis block header aggregateCommit.height must equal 0',
				);
			});

			it('should throw error if aggregateCommit.certificateSignature is not empty buffer', () => {
				const block = getGenesisBlockAttrs();
				const blockHeader = new BlockHeader({
					...block,
					aggregateCommit: {
						...block.aggregateCommit,
						certificateSignature: utils.getRandomBytes(32),
					},
				});

				expect(() => blockHeader.validateGenesis()).toThrow(
					'Genesis block header aggregateCommit.certificateSignature must be empty bytes',
				);
			});

			it('should throw error if aggregateCommit.aggregationBits is not empty buffer', () => {
				const block = getGenesisBlockAttrs();
				const blockHeader = new BlockHeader({
					...block,
					aggregateCommit: { ...block.aggregateCommit, aggregationBits: utils.getRandomBytes(32) },
				});

				expect(() => blockHeader.validateGenesis()).toThrow(
					'Genesis block header aggregateCommit.aggregationBits must be empty bytes',
				);
			});

			it('should throw error if signature is not empty buffer', () => {
				const block = getGenesisBlockAttrs();
				const blockHeader = new BlockHeader({ ...block, signature: utils.getRandomBytes(32) });

				expect(() => blockHeader.validateGenesis()).toThrow(
					'Genesis block header signature must be empty bytes',
				);
			});
		});
	});
});
