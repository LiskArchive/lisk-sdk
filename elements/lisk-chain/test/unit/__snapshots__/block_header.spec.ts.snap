// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`block_header blockHeaderSchema should be valid schema with signature but without id 1`] = `
Object {
  "$id": "/block/header/3",
  "properties": Object {
    "aggregateCommit": Object {
      "fieldNumber": 14,
      "properties": Object {
        "aggregationBits": Object {
          "dataType": "bytes",
          "fieldNumber": 2,
        },
        "certificateSignature": Object {
          "dataType": "bytes",
          "fieldNumber": 3,
        },
        "height": Object {
          "dataType": "uint32",
          "fieldNumber": 1,
        },
      },
      "required": Array [
        "height",
        "aggregationBits",
        "certificateSignature",
      ],
      "type": "object",
    },
    "assetRoot": Object {
      "dataType": "bytes",
      "fieldNumber": 7,
    },
    "eventRoot": Object {
      "dataType": "bytes",
      "fieldNumber": 8,
    },
    "generatorAddress": Object {
      "dataType": "bytes",
      "fieldNumber": 5,
      "format": "lisk32",
    },
    "height": Object {
      "dataType": "uint32",
      "fieldNumber": 3,
    },
    "impliesMaxPrevote": Object {
      "dataType": "boolean",
      "fieldNumber": 12,
    },
    "maxHeightGenerated": Object {
      "dataType": "uint32",
      "fieldNumber": 11,
    },
    "maxHeightPrevoted": Object {
      "dataType": "uint32",
      "fieldNumber": 10,
    },
    "previousBlockID": Object {
      "dataType": "bytes",
      "fieldNumber": 4,
    },
    "signature": Object {
      "dataType": "bytes",
      "fieldNumber": 15,
    },
    "stateRoot": Object {
      "dataType": "bytes",
      "fieldNumber": 9,
    },
    "timestamp": Object {
      "dataType": "uint32",
      "fieldNumber": 2,
    },
    "transactionRoot": Object {
      "dataType": "bytes",
      "fieldNumber": 6,
    },
    "validatorsHash": Object {
      "dataType": "bytes",
      "fieldNumber": 13,
    },
    "version": Object {
      "dataType": "uint32",
      "fieldNumber": 1,
    },
  },
  "required": Array [
    "version",
    "timestamp",
    "height",
    "previousBlockID",
    "generatorAddress",
    "transactionRoot",
    "assetRoot",
    "eventRoot",
    "stateRoot",
    "maxHeightPrevoted",
    "maxHeightGenerated",
    "impliesMaxPrevote",
    "validatorsHash",
    "aggregateCommit",
    "signature",
  ],
  "type": "object",
}
`;

exports[`block_header blockHeaderSchemaWithId should be valid schema with signature and id 1`] = `
Object {
  "$id": "/block/header/3",
  "properties": Object {
    "aggregateCommit": Object {
      "fieldNumber": 14,
      "properties": Object {
        "aggregationBits": Object {
          "dataType": "bytes",
          "fieldNumber": 2,
        },
        "certificateSignature": Object {
          "dataType": "bytes",
          "fieldNumber": 3,
        },
        "height": Object {
          "dataType": "uint32",
          "fieldNumber": 1,
        },
      },
      "required": Array [
        "height",
        "aggregationBits",
        "certificateSignature",
      ],
      "type": "object",
    },
    "assetRoot": Object {
      "dataType": "bytes",
      "fieldNumber": 7,
    },
    "eventRoot": Object {
      "dataType": "bytes",
      "fieldNumber": 8,
    },
    "generatorAddress": Object {
      "dataType": "bytes",
      "fieldNumber": 5,
      "format": "lisk32",
    },
    "height": Object {
      "dataType": "uint32",
      "fieldNumber": 3,
    },
    "id": Object {
      "dataType": "bytes",
      "fieldNumber": 16,
    },
    "impliesMaxPrevote": Object {
      "dataType": "boolean",
      "fieldNumber": 12,
    },
    "maxHeightGenerated": Object {
      "dataType": "uint32",
      "fieldNumber": 11,
    },
    "maxHeightPrevoted": Object {
      "dataType": "uint32",
      "fieldNumber": 10,
    },
    "previousBlockID": Object {
      "dataType": "bytes",
      "fieldNumber": 4,
    },
    "signature": Object {
      "dataType": "bytes",
      "fieldNumber": 15,
    },
    "stateRoot": Object {
      "dataType": "bytes",
      "fieldNumber": 9,
    },
    "timestamp": Object {
      "dataType": "uint32",
      "fieldNumber": 2,
    },
    "transactionRoot": Object {
      "dataType": "bytes",
      "fieldNumber": 6,
    },
    "validatorsHash": Object {
      "dataType": "bytes",
      "fieldNumber": 13,
    },
    "version": Object {
      "dataType": "uint32",
      "fieldNumber": 1,
    },
  },
  "required": Array [
    "version",
    "timestamp",
    "height",
    "previousBlockID",
    "generatorAddress",
    "transactionRoot",
    "assetRoot",
    "eventRoot",
    "stateRoot",
    "maxHeightPrevoted",
    "maxHeightGenerated",
    "impliesMaxPrevote",
    "validatorsHash",
    "aggregateCommit",
    "signature",
    "id",
  ],
  "type": "object",
}
`;

exports[`block_header signingBlockHeaderSchema should be valid schema without signature and id 1`] = `
Object {
  "$id": "/block/header/signing/3",
  "properties": Object {
    "aggregateCommit": Object {
      "fieldNumber": 14,
      "properties": Object {
        "aggregationBits": Object {
          "dataType": "bytes",
          "fieldNumber": 2,
        },
        "certificateSignature": Object {
          "dataType": "bytes",
          "fieldNumber": 3,
        },
        "height": Object {
          "dataType": "uint32",
          "fieldNumber": 1,
        },
      },
      "required": Array [
        "height",
        "aggregationBits",
        "certificateSignature",
      ],
      "type": "object",
    },
    "assetRoot": Object {
      "dataType": "bytes",
      "fieldNumber": 7,
    },
    "eventRoot": Object {
      "dataType": "bytes",
      "fieldNumber": 8,
    },
    "generatorAddress": Object {
      "dataType": "bytes",
      "fieldNumber": 5,
      "format": "lisk32",
    },
    "height": Object {
      "dataType": "uint32",
      "fieldNumber": 3,
    },
    "impliesMaxPrevote": Object {
      "dataType": "boolean",
      "fieldNumber": 12,
    },
    "maxHeightGenerated": Object {
      "dataType": "uint32",
      "fieldNumber": 11,
    },
    "maxHeightPrevoted": Object {
      "dataType": "uint32",
      "fieldNumber": 10,
    },
    "previousBlockID": Object {
      "dataType": "bytes",
      "fieldNumber": 4,
    },
    "stateRoot": Object {
      "dataType": "bytes",
      "fieldNumber": 9,
    },
    "timestamp": Object {
      "dataType": "uint32",
      "fieldNumber": 2,
    },
    "transactionRoot": Object {
      "dataType": "bytes",
      "fieldNumber": 6,
    },
    "validatorsHash": Object {
      "dataType": "bytes",
      "fieldNumber": 13,
    },
    "version": Object {
      "dataType": "uint32",
      "fieldNumber": 1,
    },
  },
  "required": Array [
    "version",
    "timestamp",
    "height",
    "previousBlockID",
    "generatorAddress",
    "transactionRoot",
    "assetRoot",
    "eventRoot",
    "stateRoot",
    "maxHeightPrevoted",
    "maxHeightGenerated",
    "impliesMaxPrevote",
    "validatorsHash",
    "aggregateCommit",
  ],
  "type": "object",
}
`;
